### DISCUSSION TITLE
Server Sent Events Question

### DISCUSSION THREAD

**Discussion by dbrucknr** [encoredev/encore#1676] [Help]
I wanted to reach out to the Encore community and see if there was any support for Server Sent Events from an encore.ts backend?

I was trying to dig through the source code and came up with an approach like this:

```
export const test = api.raw(
  { method: "GET", path: "/subscribe", expose: true },
  async (req, res) => {
    res.setHeader("Content-Type", "text/event-stream");
    res.setHeader("Cache-Control", "no-cache");
    res.setHeader("Connection", "keep-alive");
    res.flushHeaders();

    const sendEvent = () =>
      res.write(`data: ${JSON.stringify({ message: "Hello, client!" })}\n\n`);

    const interval = setInterval(sendEvent, 5000);

    req.on("close", () => {
      clearInterval(interval);
    });
  }
);
```

**Comment by marcuskohlberg** [comment]
Could you not use e.g. a cron job for this? https://encore.dev/docs/ts/primitives/cron-jobs

**Reply by dbrucknr** [reply]
Hey there Marcus,

Many thanks for the reply!

May I ask a bit more on how the cron-job might help me subscribe external services and clients to background tasks, jobs, and even requests for data to the encore backend?

There's a reasonable chance I've misunderstood the entire functionality of the cron job api in encore. If I may explain further, what I had wondered about with regards to [Server Sent Events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events) might be akin to encore's streaming endpoints. 

I looked at the documentation here [Streaming APIs](https://encore.dev/docs/ts/primitives/streaming-apis)

I think the streaming endpoints are close to what I had originally wanted to achieve, but I experienced some difficulty in getting one to work with the JavaScript EventSource API. Inferring from the docs, it looked like these endpoints provided a full duplex Websocket tunnel to transmit data between services. 

The type of development task I was interested in achieving was with the simplex communication type provided by the Server Sent Events protocol. In my case I wanted my clients, scripts, and other codebases to subscribe to the encore.ts backend, which would then forward events captured in another service, like Redis / Memcache / Kafka / NATS / etc. through the subscription tunnel in real-time. 

Let me know your thoughts, or if I've crossed wires / misunderstood anything.

Also, many thanks for encore - I'm very excited to see where the library goes!

**Reply by marcuskohlberg** [reply]
Hey, I probably inferred incorrectly what you were looking to do.
Cron jobs are primarily useful for scheduling recurring/periodic tasks. However, they don’t directly handle real-time events like you are looking for.

Regarding Streaming APIs, would the [streamOut](https://encore.dev/docs/ts/primitives/streaming-apis#streamout) variety not be suitable for your use case? It is intended for streaming from server to client.

**Reply by dbrucknr** [reply]
Hey, sorry for the slow response back on this. 

I think the streamOut API is fine for cases when you want / need a `WebSocket`, but, I'm more interested in knowing if encore had native Server Sent Event (i.e `EventSource`) support, as I don't need a full duplex communication pattern for what I'm looking to do (Pub / Sub clients to an endpoint). My clients won't be transmitting anything into the encore backend, which is why I don't need the duplex communication protocol, but a more robust simplex communication protocol.

I can get Server Sent Events working just fine using the sample code I posted in my question - which sets the `text/event-stream` header, but I wanted to reach out and ask if there was a `encore` EventSource streaming API.

It sounds like at this time there isn't one, which is fine, as I think writing the `api.raw` method works well.
